### ISSUE TITLE
why not allow effects that do not return actions

### ISSUE THREAD

**Issue by bdwain** [raisemarketplace/redux-loop#54]
I know in the docs it explicitly states that you have to return an action. But would it break the elm architecure if you didn't? That's bascially what none() does. I just tried it and it's trivial to implement. 

I can see a lot of scenarios where it makes sense to me to do this, and a subscriber would just be weird. For example, i have an object that can get updated, and when it does I want to trigger a write to localstorage. the new state depends on the previous state, so the reducer needs to be involved before the write to localstorage. but after that, there's nothing to do. i would like to return from my reducer something like

`loop(newState, Effects.call(writeAuthUpdatesToStorage, newState));`

instead of having to subscribe to the store and compare newUser and oldUser every state change.

Is this a change that could be considered? (I'd make the pull request if so)

**Comment by joeycadle** [comment]
+1

**Comment by lukewestby** [comment]
The way this is usually handled in Elm, especially as of 0.17, is to have a `NoOp` action which effectively does nothing and just returns the state and a `none` effect. You can use the same pattern with redux-loop, and you don't even need to handle this noop action in your reducer -- you can just ignore it completely. This should accomplish what you need without having to special-case effects that don't need to do anything afterword on the library side.

**Comment by bdwain** [comment]
That definitely works. The main reason I preferred the ability to return nothing was that if I was working with redux dev tools or something, adding a bunch of noop actions seems like it's just polluting the history of my state transitions and adds noise when trying to debug things. And it was super simple to implement. Basically just added .filter(Boolean) to the resolved promise array in effectToPromie (see what i did [here](https://github.com/bdwain/redux-loop/commit/3545306e71afb08a1260cc7cfa07b97baa216e3b)).

So basically it just seemed like an appropriate diversion from elm considering the tools available in redux and the development patterns I've had.

**Comment by bdwain** [comment]
After using redux-loop a bit more in practice, I've started to have some thoughts on how the api could encourage a little more code reuse and simplify a few things.

One issue with forcing effects to return/resolve with redux actions is that the implementations of the side-effects become tied to redux/redux-loop. Suddenly a pretty standard function to make a request, such as this, 

```js
import apiClient from './apiClient.js'; //to abstract away fetch details like json parsing and error codes

function fetchItems(userid){
  return apiClient(`foo.com/users/${userId}/items`)
      .then(rsp => {
           //process rsp
           //return processed result
       }).catch(err => {
            //maybe do something?
       });
}
```

can't be used by redux-loop without changing it in a way that makes it unusable by something else (you can no longer reject the promise, and your resolution needs to be a redux action and not just the list of items). 

Additionally, even if you wanted it only for redux-loop, it would be impossible (or at least awkward) to use a side-effect to return different types of actions depending on how it is called (I haven't had a need to do this yet, but it's not hard to imagine a case where you'd want it)

I know this goes even further than my original suggestion of allowing side-effects to not return actions, but what if instead, we allowed them to become completely agnostic to redux-loop/redux in general, and instead return whatever they felt was necessary to complete their purpose (so for async methods, prob would be a promise that resolves or rejects to indicate success).

So instead of defining the actions triggered (or not) by the effect in the method, it could be defined in the Effect object itself.

The new effect objects returned by a reducer might look like this
```js

//actions.js
function itemsLoadedActionCreator(items){
    return {
       type: 'ITEMS_LOADED',
       items: items
    };
}
// ... other action creators 

//reducer.js
return loop(newState, Effect.promise(fetchItems, itemsLoadedActionCreator, failHandler, params...));

Effect.call(writeToLocalStore, syncMethodActionCreator, params...)

//Effects.constant could be implemented as syntactic sugar for 
Effects.call(null, constantAction());

//batch would be basically the same
//I imagine lift would be similar, though I haven't looked into it much yet
```

And now your effects are completely reusable and agnostic to your choice of state management. You could swap out redux-loop or even redux iteself without having to change them.

Let me know if you want me to move this to a new issue. I started writing it as a follow up to the original question, but at this point, whether or not the handlers are optional or required (which is basically what the oriiginal question boils down to) seems to be a separate question from whether or not this all makes sense.

**Comment by lukewestby** [comment]
@bdwain yes, let's please start a new issue about making this change

**Comment by bdwain** [comment]
Sounds good. I made #66. Hopefully that will make it easy to implement my original suggestion of allowing side-effects to not trigger an action without having to special case anything. I'll close this for now.
